### PR TITLE
[DSIP-78][Data Quality] Remove data quality module

### LINKED ISSUE #16728: [DSIP-78][Data Quality] Suggest remove data quality module
### Search before asking

- [X] I had searched in the [DSIP](https://github.com/apache/dolphinscheduler/issues/14102) and found no similar DSIP.

### Motivation

The current `data quality`  task type can hardly be used normally. Since version 3.2.0, this task type has been unable to be used normally, which is equivalent to leaving it vacant.

I search in the issue list found that `data quality` task has a lot of bug issue, and no one maintains them, and there are also a lot of CVEs.

Most importantly, `data quality` is seriously coupled to the current code base. So that dependencies can't be optimized, binary package size can't be reduced, and code maintenance cost is extremely high.

So I suggest remove it.

### Design Detail

_No response_

### Compatibility, Deprecation, and Migration Plan

_No response_

### Test Plan

_No response_

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by SbloodyS**
## Purpose of the pull request

close #16728

## Verify this pull request

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

(or)

## Pull Request Notice
[Pull Request Notice](https://github.com/apache/dolphinscheduler/blob/dev/docs/docs/en/contribute/join/pull-request.md)

If your pull request contain incompatible change, you should also add it to `docs/docs/en/guide/upgrede/incompatible.md`

### CODE REVIEW

**Comment by davidzollo** [comment]
-1，there're no conclusion for now, please don't do this op

**Comment by SbloodyS** [comment]
> -1，there're no conclusion for now, please don't do this op

Based on the discussion of issue #16728, we can draw the following conclusions.

1. More than half of the maintainers agreed.
2. Half of the users who disagree don't understand the difference between DS and Hadoop and data-quality. It's meaningless.
3. The remaining half users want to refactor the data-quality module instead of removing it.

No one among the maintainers on the current community is willing to refactor this module. And data-quality module has seriously blocked the progress of #16098 which is is very important for the next version's release. So I think the conclusion of removal is obvious.

**Comment by davidzollo** [comment]
> > -1，there're no conclusion for now, please don't do this op
> 
> Based on the discussion of issue #16728, we can draw the following conclusions.
> 
> 1. More than half of the maintainers agreed.
> 2. Half of the users who disagree don't understand the difference between DS and Hadoop and data-quality. It's meaningless.
> 3. The remaining half users want to refactor the data-quality module instead of removing it.
> 
> No one among the maintainers on the current community is willing to refactor this module. And data-quality module has seriously blocked the progress of #16098 which is is very important for the next version's release. So I think the conclusion of removal is obvious.


Apache emphasizes achieving consensus. If there are significant differences and we cannot be resolved in the short term, we may choose to temporarily shelve the proposal, allowing much more time for more information before re-discussing. 

Considering this is a major decision, I suggest you send a vote email to the dev mailing list.

**Comment by SbloodyS** [comment]
> Apache emphasizes achieving consensus.

First of all, the current consensus does not require all people to reach an agreement, but only more than half. For more info you can take a look at https://github.com/apache/comdev-site/pull/189

> If there are significant differences and we cannot be resolved in the short term, we may choose to temporarily shelve the proposal, allowing much more time for more information before re-discussing.

Like I said, no one among the maintainers on the current community is willing to refactor this module. How to achieve this without anyone willing to take responsibility?

> Considering this is a major decision, I suggest you send a vote email to the dev mailing list.

I think issue and dev mail list have the same meaning on this issue. And all active PMC/Committer have been included in this issue.

**Comment by davidzollo** [comment]
> > Apache emphasizes achieving consensus.
> 
> First of all, the current consensus does not require all people to reach an agreement, but only more than half. For more info you can take a look at [apache/comdev-site#189](https://github.com/apache/comdev-site/pull/189)
> 
> > If there are significant differences and we cannot be resolved in the short term, we may choose to temporarily shelve the proposal, allowing much more time for more information before re-discussing.
> 
> Like I said, no one among the maintainers on the current community is willing to refactor this module. How to achieve this without anyone willing to take responsibility?
> 
> > Considering this is a major decision, I suggest you send a vote email to the dev mailing list.
> 
> I think issue and dev mail list have the same meaning on this issue. And all active PMC/Committer have been included in this issue.

Github Issue can't instead of mail, especially for big event.
I think we can try to find some contributors who'd like to refactor this module if this function is finally kept.

By the way, i don't disagree with removing this module;  I’m just concerned about the significant impact it will have on users. What I see is that many users want to keep this module, while the maintainers are inclined to remove it to make refactoring easier. Therefore, this decision requires great caution. If a vote does not take place in the dev mailing list, then it should happen in private, not in an issue thread.

**Comment by SbloodyS** [comment]
> Github Issue can't instead of mail, especially for big event. I think we can try to find some contributors who'd like to refactor this module if this function is finally kept.
> 
> By the way, i don't disagree with removing this module; I’m just concerned about the significant impact it will have on users. What I see is that many users want to keep this module, while the maintainers are inclined to remove it to make refactoring easier. Therefore, this decision requires great caution. If a vote does not take place in the dev mailing list, then it should happen in private, not in an issue thread.

I don't think this module is a big event. This feature was introduced in PR #4830 Since Feb 21, 2021 without any mailing or github issue discussion. For more than three years, apart from this PR author, no contributor has contributed to this function, and there are endless issue of bugs and improvement, and there is no substantial code change in this function. This author also doesn't want to maintain this function and vote +1 for removal.

**Comment by davidzollo** [comment]
At that time, there was no DSIP mechanism in place, and much of the communication took place during community meetings. The feature I’m referring to took 8-9 months to develop and implement. If this is not a major feature, I believe that statement would be inaccurate.

**Comment by SbloodyS** [comment]
> At that time, there was no DSIP mechanism in place, and much of the communication took place during community meetings. The feature I’m referring to took 8-9 months to develop and implement. If this is not a major feature, I believe that statement would be inaccurate.

I don't agree with this opinion. Since dolphinscheduler is focusing the modern data orchestration platform. Data-Quality is focusing accuracy and consistency of data. The relationship between two of them is equivalent to Flink and Flink-CDC. There are many mature examples at present.

**Review by ruanwenjun — COMMENTED** [review]
We should add ddl to drop the dq table.

**Comment by SbloodyS** [comment]
> We should add ddl to drop the dq table.

We can remove the create table ddl in the init sql and give some drop table ddl to users in docs to let users decide whether or not to execute it instead of execute it by default. WDYT?

**Comment by zhongjiajie** [comment]
> > We should add ddl to drop the dq table.
> 
> We can remove the create table ddl in the init sql and give some drop table ddl to users in docs to let users decide whether or not to execute it instead of execute it by default. WDYT?

How about giving entry point script to remove exists table and package it into binary tarball instead of document, cause we support only two of databases in prod, and we should keep thing easy to use

**Comment by SbloodyS** [comment]
> > > We should add ddl to drop the dq table.
> > 
> > 
> > We can remove the create table ddl in the init sql and give some drop table ddl to users in docs to let users decide whether or not to execute it instead of execute it by default. WDYT?
> 
> How about giving entry point script to remove exists table and package it into binary tarball instead of document, cause we support only two of databases in prod, and we should keep thing easy to use

I think it will be more traversal for some users to provide users with operating DLLs intuitively since this is a dangerous operation for drop table...

**Comment by zhongjiajie** [comment]
For proposal only, it seems inconsistency exists, although some of the PMCs in #16728 already agreed to remove it, David is right and he could challenge it and ask to vote in the dev mailing list. So maybe should vote in dev mail thread, And I think the vote result should be the **ONLY** result of this PR continue or not instead of personal emotions.
I found out author @zixi0825 also voted +1 in https://github.com/apache/dolphinscheduler/issues/16728#issuecomment-2434073450, and @qingwli, who has maintained data quality for some time, voted -1 for removing it. 

And for this feature, I think it should better act as a plugin instead of a built-in function. Especially since it has many bugs and CVEs and not team member want to maintain it. So I personally would vote +1 for removing it. 

And BTW, the time cost for the feature development should not as a standard to measure the importance or not.

**Comment by zhongjiajie** [comment]
> > > > We should add ddl to drop the dq table.
> > > 
> > > 
> > > We can remove the create table ddl in the init sql and give some drop table ddl to users in docs to let users decide whether or not to execute it instead of execute it by default. WDYT?
> > 
> > 
> > How about giving entry point script to remove exists table and package it into binary tarball instead of document, cause we support only two of databases in prod, and we should keep thing easy to use
> 
> I think it will be more traversal for some users to provide users with operating DLLs intuitively since this is a dangerous operation for drop table...

I mean we should add new bash script like https://github.com/apache/dolphinscheduler/blob/dev/dolphinscheduler-tools/src/main/bin/migrate-lineage.sh which is separated form `upgrade-schema.sh`, for easy use of end users. But it not important thing, the exists data should be keep and we can provide document to notice our users

**Comment by SbloodyS** [comment]
> I mean we should add new bash script like https://github.com/apache/dolphinscheduler/blob/dev/dolphinscheduler-tools/src/main/bin/migrate-lineage.sh which is separated form upgrade-schema.sh, for easy use of end users. But it not important thing, the exists data should be keep and we can provide document to notice our users

Ok.

**Comment by SbloodyS** [comment]
> For proposal only, it seems inconsistency exists, although some of the PMCs in #16728 already agreed to remove it, David is right and he could challenge it and ask to vote in the dev mailing list. So maybe should vote in dev mail thread, And I think the vote result should be the **ONLY** result of this PR continue or not instead of personal emotions. I found out author @zixi0825 also voted +1 in [#16728 (comment)](https://github.com/apache/dolphinscheduler/issues/16728#issuecomment-2434073450), and @qingwli, who has maintained data quality for some time, voted -1 for removing it.
> 
> And for this feature, I think it should better act as a plugin instead of a built-in function. Especially since it has many bugs and CVEs and not team member want to maintain it. So I personally would vote +1 for removing it.
> 
> And BTW, the time cost for the feature development should not as a standard to measure the importance or not.

I will raise a vote in dev mail list.

**Comment by zhongjiajie** [comment]
FYI, vote mail in https://lists.apache.org/thread/0tldm33skkbrfgbt01bvd610z5zmb725

**Comment by qingwli** [comment]
I voted -1 for removing Data quality before, but the author gave us another option about the plugin instead of the built-in function. I think it's a good way. 

Actually, ds's data quality is not an out-of-the-box  module. The author builds a framework but still needs lots of work that can be used and used simply. The author has another github repo about the greater and more functional data quality project. If we enhance the data quality in DS, it means we will do the same work.

So I agree to remove the data quality module in ds, and use a plugin to let users continue to use the better data quality check functions.

About removing tables in ds, I agree @zhongjiajie new bash scrip way.

**Comment by fuchanghai** [comment]
> I voted -1 for removing Data quality before, but the author gave us another option about the plugin instead of the built-in function. I think it's a good way.
> 
> Actually, ds's data quality is not an out-of-the-box module. The author builds a framework but still needs lots of work that can be used and used simply. The author has another github repo about the greater and more functional data quality project. If we enhance the data quality in DS, it means we will do the same work.
> 
> So I agree to remove the data quality module in ds, and use a plugin to let users continue to use the better data quality check functions.
> 
> About removing tables in ds, I agree @zhongjiajie new bash scrip way.

agree .

**Comment by Gallardot** [comment]
> I voted -1 for removing Data quality before, but the author gave us another option about the plugin instead of the built-in function. I think it's a good way.
> 
> Actually, ds's data quality is not an out-of-the-box module. The author builds a framework but still needs lots of work that can be used and used simply. The author has another github repo about the greater and more functional data quality project. If we enhance the data quality in DS, it means we will do the same work.
> 
> So I agree to remove the data quality module in ds, and use a plugin to let users continue to use the better data quality check functions.
> 
> About removing tables in ds, I agree @zhongjiajie new bash scrip way.

Agree. I voted +1.
Dolphinscheduler should focus on workflow orchestration and task scheduling. It should then provide a good plugin mechanism to integrate with other projects, thereby building a better ecosystem. Not just data quality, I believe Flink real-time stream processing tasks should also be removed from dolphinscheduler.

**Comment by qingwli** [comment]
About the Flink real-time stream processing task, we have rewritten it for our platform. And our platform looks like a big data schedule system. @davidzollo told want to have a quick meet about the Real-time part in Ds, We have finished some doc about introducing our Flink, Do you have some time so we can discuss this part? CC @Gallardot

**Comment by SbloodyS** [comment]
> I mean we should add new bash script like https://github.com/apache/dolphinscheduler/blob/dev/dolphinscheduler-tools/src/main/bin/migrate-lineage.sh which is separated form upgrade-schema.sh, for easy use of end users. But it not important thing, the exists data should be keep and we can provide document to notice our users

We remove data-quality related table in init sql but not add it in upgrade sql. It will cause the schema check in CI failed... @zhongjiajie @ruanwenjun

File: dolphinscheduler-common/src/test/resources/sql/mysql_dml.sql [hunk 1]
```diff
@@ -38,13 +38,6 @@ DROP PROCEDURE dolphin_t_ds_tenant_insert_default;
...
 
--- data quality support choose database
-INSERT IGNORE INTO `t_ds_dq_rule_input_entry`
-(`id`, `field`, `type`, `title`, `value`, `options`, `placeholder`, `option_source_type`, `value_type`, `input_type`, `is_show`, `can_edit`, `is_emit`, `is_validate`, `create_time`, `update_time`)
-VALUES(30, 'src_database', 'select', '$t(src_database)', NULL, NULL, 'please select source database', 0, 0, 0, 1, 1, 1, 1, current_timestamp, current_timestamp);
-INSERT IGNORE INTO `t_ds_dq_rule_input_entry`
-(`id`, `field`, `type`, `title`, `value`, `options`, `placeholder`, `option_source_type`, `value_type`, `input_type`, `is_show`, `can_edit`, `is_emit`, `is_validate`, `create_time`, `update_time`)
-VALUES(31, 'target_database', 'select', '$t(target_database)', NULL, NULL, 'please select target database', 0, 0, 0, 1, 1, 1, 1, current_timestamp, current_timestamp);
 
 INSERT IGNORE INTO `t_ds_relation_rule_input_entry`
```
Comment (ruanwenjun) [inline dolphinscheduler-common/src/test/resources/sql/mysql_dml.sql]: This also need to remove

File: dolphinscheduler-dao/src/main/resources/sql/dolphinscheduler_h2.sql
Comment (ruanwenjun) [inline dolphinscheduler-dao/src/main/resources/sql/dolphinscheduler_h2.sql:1]: Please rm t_ds_relation_rule_execute_sql

**Review by ruanwenjun — CHANGES_REQUESTED** [review]

**Review by ruanwenjun — APPROVED** [review]
LGTM

**Review by zhongjiajie — APPROVED** [review]
LGTM, thanks

**Pull request merged**